### PR TITLE
SOLR-17321: Bump minimum required Java version to 21

### PR DESCRIPTION
**Pull request by iamsanjay**
https://issues.apache.org/jira/browse/SOLR-17321

We are upgrading the minimum Java version for Solr main branch to 21. However, at the same, It has been suggested to be not so aggressive with SolrJ (and thus solr-api, a dependency) Java version – setting it to 17.

**Depreciated API**
1.  AccessController [Deprecated since 17 and marked for removal]
2. AccessControlContext [Deprecated since 17 and marked for removal]
3. java.net.URL ctors [Deprecated since 20]
4. SecurityManager [Deprecated since 17 and marked for removal]
5. Thread#getId() [Deprecated in Java 19]

To not make any extra code changes below annotation were used:

```
 @SuppressForbidden(reason = "<custom message>") //Still need to decide what the message should be for all. Thinking of adding "Deprecated in Java <version>"

@SuppressWarnings("removal")
```

However, at few places, for demonstration purpose, URL has been created via new suggested way instead of using the ctors:

```
url = new URL(urlStr);
url = URI.create(urlStr).toURL();
```
Or,
 ```
URL solrURL = new URI(solrUrl).toURL();
```

Also, Bump up the ecj version to 3.37 and ca.cutterslade.analyze to 1.9.2.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Comment by iamsanjay** [comment]
For now two things pending
1. Set different Java level for SolrJ and Solr. Currently both are 21.
2. And, another one is to change the test-via-crave.yml file to run the build with Java 21.

Feel free to push changes to this branch.

File: solr/core/src/test/org/apache/solr/response/transform/TestSubQueryTransformerDistrib.java [hunk 1]
```diff
@@ -134,6 +135,7 @@ public static void setupCluster() throws Exception {
   }
 
   @SuppressWarnings("serial")
+  @SuppressForbidden(reason = "java.net.URL ctors deprecated since Java 20")
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/response/transform/TestSubQueryTransformerDistrib.java:138]: You fix some and suppress others.  Is there a pattern/approach to it?  (just curious; maybe not)

**Review by dsmiley — COMMENTED** [review]
RE URL:  If you have IntelliJ Ultimate edition (not sure if Community), you could do "Structural Replace" to find & replace systematically everywhere.

FYI as an ASF committer, you can get that edition for free -- https://www.jetbrains.com/shop/eform/apache?product=ALL

**Comment by dsmiley** [comment]
> @SuppressForbidden(reason = "<custom message>") //Still need to decide what the message should be for all. Thinking of adding "Deprecated in Java <version>"

Ideally those are addressed for real, otherwise they will continue to linger.  Could be a separate JIRA/PR.

File: build.gradle [hunk 2]
```diff
@@ -31,8 +31,8 @@ plugins {
 }
 
 // Declare default Java versions for the entire project and for SolrJ separately
-rootProject.ext.minJavaVersionDefault = JavaVersion.VERSION_11
-rootProject.ext.minJavaVersionSolrJ = JavaVersion.VERSION_11
+rootProject.ext.minJavaVersionDefault = JavaVersion.VERSION_21
+rootProject.ext.minJavaVersionSolrJ = JavaVersion.VERSION_21
```
Comment (janhoy) [inline build.gradle:35]: Use Java 17 for SolrJ

File: solr/core/src/java/org/apache/solr/cli/PostTool.java [hunk 3]
```diff
@@ -601,6 +604,7 @@ protected static String normalizeUrlEnding(String link) {
    * @param out output stream to write to
    * @return number of pages crawled on this level and below
    */
+  @SuppressForbidden(reason = "java.net.URL ctors deprecated since Java 20")
```
Comment (epugh) [inline solr/core/src/java/org/apache/solr/cli/PostTool.java:607]: Do we need this suppress?

File: solr/core/src/java/org/apache/solr/cli/PostTool.java [hunk 4]
```diff
@@ -624,13 +628,14 @@ protected int webCrawl(int level, OutputStream out) {
...
           url = (result.redirectUrl != null) ? result.redirectUrl : url;
           URL postUrl =
-              new URL(
-                  appendParam(
-                      solrUpdateUrl.toString(),
-                      "literal.id="
-                          + URLEncoder.encode(url.toString(), UTF_8)
-                          + "&literal.url="
-                          + URLEncoder.encode(url.toString(), UTF_8)));
+              new URI(
```
Comment (epugh) [inline solr/core/src/java/org/apache/solr/cli/PostTool.java:631]: If you run tidy do some white space changes go away?

File: solr/core/src/java/org/apache/solr/cli/PostTool.java [hunk 5]
```diff
@@ -815,7 +820,7 @@ public void postFile(File file, OutputStream output, String type) throws Malform
           urlStr =
               appendParam(urlStr, "literal.id=" + URLEncoder.encode(file.getAbsolutePath(), UTF_8));
         }
-        url = new URL(urlStr);
+        url = URI.create(urlStr).toURL();
```
Comment (epugh) [inline solr/core/src/java/org/apache/solr/cli/PostTool.java:823]: Curious if using a URI would be better than constantly converting back to string?   Maybe another pr...

**Review by epugh — COMMENTED** [review]
Great to see movement.   I would prefer to see the Java deprecations fixed versus more suppress annotations!!!   Excited for this.

**Comment by iamsanjay** [comment]
> RE URL: If you have IntelliJ Ultimate edition (not sure if Community), you could do "Structural Replace" to find & replace systematically everywhere.
Like always, Thanks for the air support. "Structural Replace" is a Game Changer!

**Comment by iamsanjay** [comment]
> Would it be too much to ask to separate all this URL conversion stuff to a separate PR, ultimately being its own commit focused on exactly that? Needn't be another JIRA; needn't touch CHANGES.txt in that commit. If I had to do this, I'd start with the trick of taking a GitHub PR URL and adding the ".patch" (or maybe it's ".diff" for a different format) and then copy-paste into IntellIJ "Apply patch from Clipboard" and selectively exclude stuff unrelated at that point (or after).

So SuppressForbidden is okay for now, and later on (Just after this one!) deal with removing it in a different PR?

**Comment by dsmiley** [comment]
I suggest flipping the order.  Prep the code for Java 21 compatibility
first so that we don't go adding SupressForbidden to many classes to only
then fix the issue later.  i.e. visit these classes once not twice.  You
know exactly how to fix it and have even starting doing so; it's not a
tricky matter.  Again, IntelliJ will do much of the work.

**Comment by iamsanjay** [comment]
Test case failing, on main branch as well!
```
./gradlew :solr:core:test --tests "org.apache.solr.search.function.TestDenseVectorFunctionQuery.floatFieldVectors_shouldReturnFloatSimilarity" -Ptests.jvms=96 "-Ptests.jvmargs=-XX:TieredStopAtLevel=1 -XX:+UseParallelGC -XX:ActiveProcessorCount=1 -XX:ReservedCodeCacheSize=120m" -Ptests.seed=8F0CED8E85B304BD -Ptests.timeoutSuite=600000! -Ptests.file.encoding=US-ASCII
```

**Comment by iamsanjay** [comment]
Setting different Java version for main and SolrJ resulting in error.  Currently, `:solr:api:ecjLintTest ` task is failing and below is the error.

```
> Task :solr:api:ecjLintTest
----------
1. ERROR in /Users/sanjaydutt/Documents/Solr/solr-main/solr/solr/api/src/test/org/apache/solr/util/TestSolrVersion.java (at line 19)
        import org.apache.solr.SolrTestCase;
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The import org.apache.solr.SolrTestCase cannot be resolved
----------
2. ERROR in /Users/sanjaydutt/Documents/Solr/solr-main/solr/solr/api/src/test/org/apache/solr/util/TestSolrVersion.java (at line 23)
        public class TestSolrVersion extends SolrTestCase {
                                             ^^^^^^^^^^^^
SolrTestCase cannot be resolved to a type

```

The api test cases not able to access the class, `SolrTestClass`, from the main branch.  Further, I logged the java version that has been used to configure SolrJ via adding println statements.  

`ecj-lint.gradle`

```
allprojects {
  plugins.withType(JavaPlugin) {
    // Create a [sourceSetName]EcjLint task for each source set
    // with a non-empty java.srcDirs. These tasks are then
    // attached to project's "ecjLint" task.

    println "The project name is " + project + " and the sourceCompatibility is " + project.sourceCompatibility

    def lintTasks = sourceSets.collect { sourceSet ->
      def srcDirs = sourceSet.java.srcDirs.findAll { dir -> dir.exists() }

      tasks.create(sourceSet.getTaskName("ecjLint", null), JavaExec, {task ->
....
...
```

**Output**

```
> Configure project :solr:server
The project name is project ':solr:server' and the sourceCompatibility is 21

> Configure project :solr:core
The project name is project ':solr:core' and the sourceCompatibility is 21

> Configure project :solr:api
The project name is project ':solr:api' and the sourceCompatibility is 17

> Configure project :solr:solrj
The project name is project ':solr:solrj' and the sourceCompatibility is 17
```

The version has changed for SolrJ and for api as well. However, as @dsmiley pointed out before that test classes for both of these project should point to 21. Still looking for a way How we can achieve that.

**Pull request closed without merging**